### PR TITLE
fix: Remove stale metrics file on start-up

### PR DESCRIPTION
**Pull request by Fiona-Waters**
**What this PR does / why we need it**:

This PR adds logic to cleanup stale metrics files on job startup. This is for training hub trainer only. Training hub trainer does not support pause/resume functionailty yet and if the job restarts the metrics file will be appended but will repeat step 0 etc. Therefore removing any existing metrics files at the beginning makes most sense in order to ensure correct progression tracking in the UI.

If/When pause/resume functionality is added later this will need to be updated.

I've tested this manually on my teams OpenShift cluster.

**Which issue(s) this PR fixes** _(optional, in `Fixes #<issue number>, #<issue number>, ...` format, will close the issue(s) when PR gets merged)_:

Fixes # https://issues.redhat.com/browse/RHOAIENG-41647

**Checklist:**

- [ ] [Docs](https://www.kubeflow.org/docs/) included if any changes are user facing

## Summary by CodeRabbit

* **New Features**
  * Metadata-driven algorithm support enabling dynamic metric detection and more robust termination/reporting across training variants.

* **Bug Fixes**
  * Startup now removes stale training metrics from previous runs (performed only by the primary pod), logs cleanup activity, and still starts the metrics server even if cleanup fails.
  * Termination metric extraction is more resilient to differing metric key names.

* **Tests**
  * Added tests for metrics cleanup, metadata handling, primary-pod detection, and related edge cases.

### CODE REVIEW

File: kubeflow/trainer/rhai/traininghub.py [hunk 1]
```diff
@@ -491,6 +491,47 @@ def log_message(self, format, *args):
...
+                        filename = os.path.basename(file_path)
+                        print(f"[Kubeflow] Warning: Could not remove {filename}: {e}", flush=True)
+
+            if files_removed > 0:
+                file_word = "files" if files_removed != 1 else "file"
+                print(f"[Kubeflow] Cleaned {files_removed} stale metrics {file_word}", flush=True)
+            else:
+                print("[Kubeflow] No stale metrics files found", flush=True)
+
+        except OSError as e:
```
[the remark this answers is not in the document]
Comment (Fiona-Waters) [reply kubeflow/trainer/rhai/traininghub.py]: This is the exact opposite of what coderabbit suggested

File: kubeflow/trainer/rhai/traininghub.py [hunk 2]
```diff
@@ -491,6 +491,47 @@ def log_message(self, format, *args):
...
+            print("[Kubeflow] Cleaning stale metrics files", flush=True)
+
+            # Determine file patterns based on algorithm
+            if algorithm == "sft":
+                patterns = [SFT_METRICS_FILE_PATTERN]
+            elif algorithm == "osft":
+                # All ranks
+                patterns = [
+                    OSFT_METRICS_FILE_RANK0.replace("_0.jsonl", "_*.jsonl"),
+                ]
```
[the remark this answers is not in the document]
Comment (Fiona-Waters) [reply kubeflow/trainer/rhai/traininghub.py]: I think this is outside the bounds of this PR.

File: kubeflow/trainer/rhai/traininghub_test.py [hunk 3]
```diff
@@ -690,5 +690,122 @@ def test_instrumentation_wrapper_flush_all_prints():
...
+def test_instrumentation_cleans_osft_metrics_on_startup(tmp_path):
+    """Test that OSFT metrics files are cleaned before training starts."""
+    print("Executing test: OSFT metrics cleanup on startup")
+
+    # Create stale metrics files
+    ckpt_dir = tmp_path / "checkpoints"
+    ckpt_dir.mkdir()
+
+    stale_metrics_file = ckpt_dir / "training_metrics_0.jsonl"
+    stale_metrics_file.write_text('{"step": 50, "loss": 1.2}\n')
```
[the remark this answers is not in the document]
Comment (Fiona-Waters) [reply kubeflow/trainer/rhai/traininghub_test.py]: I don't think this is necessary

File: kubeflow/trainer/rhai/traininghub.py [hunk 4]
```diff
@@ -491,6 +491,47 @@ def log_message(self, format, *args):
 
     def apply_progression_tracking():
         """Start HTTP server for metrics in background thread."""
+        # Clean stale metrics files from previous runs
+        # Since backends restart training from step 0 without checkpoint resumption,
+        # old metrics files should be deleted to avoid showing stale/incorrect progress
```
[the remark this answers is not in the document]
Comment (Fiona-Waters) [reply kubeflow/trainer/rhai/traininghub.py:493]: I like this idea.

File: kubeflow/trainer/rhai/traininghub.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (Fiona-Waters) [reply kubeflow/trainer/rhai/traininghub.py]: Update this.

File: kubeflow/trainer/rhai/traininghub_test.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (Fiona-Waters) [reply kubeflow/trainer/rhai/traininghub_test.py]: Updated this.

File: kubeflow/trainer/rhai/traininghub.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (briangallagher) [reply kubeflow/trainer/rhai/traininghub.py:493]: yep, agree

File: kubeflow/trainer/rhai/traininghub.py [hunk 5]
```diff
@@ -491,6 +492,59 @@ def log_message(self, format, *args):
...
+        pod_name_env = os.environ.get("POD_NAME")
+        pod_name = pod_name_env or hostname
+        is_primary_pod = pod_name_env is None or pod_name.endswith("-node-0")
+
+        if is_primary_pod:
+            try:
+                print("[Kubeflow] Primary pod cleaning stale metrics files", flush=True)
+
+                # Determine file patterns based on algorithm
+                if algorithm == "sft":
```
Comment (briangallagher) [inline kubeflow/trainer/rhai/traininghub.py]: Can we use class TrainingHubAlgorithms(Enum) and also include Lora?

File: kubeflow/trainer/rhai/traininghub.py [hunk 6]
```diff
@@ -491,6 +492,59 @@ def log_message(self, format, *args):
 
     def apply_progression_tracking():
         """Start HTTP server for metrics in background thread."""
+        # Clean stale metrics files from previous runs
+        # Since backends restart training from step 0 without checkpoint resumption,
+        # old metrics files should be deleted to avoid showing stale/incorrect progress
+        # Only the primary pod (rank 0, identified by -node-0 suffix in hostname) performs cleanup
+        # If pod name cannot be determined, assume primary pod (single pod or test environment)
+        hostname = os.environ.get("HOSTNAME", "")
```
Comment (briangallagher) [inline kubeflow/trainer/rhai/traininghub.py]: @Fiona-Waters Maybe we could use this opportunity to create a generic function for this logic as it seems like it could be common across the sdk. Also, I wonder is it more accurate to use JOB_COMPLETION_INDEX and PET_NODE_RANK as determining values: Add to utils.py something like: 


```
def is_primary_pod() -> bool:
    """Return True if this pod is the primary training pod.

    Detection precedence:
    1) JOB_COMPLETION_INDEX (downward API for label batch.kubernetes.io/job-completion-index)
       - Primary pod is index "0"
    2) PET_NODE_RANK (set by Torch runtime plugin)
       - Primary pod is rank "0"

    If neither signal is present, returns False (conservative).
    """
    job_index = os.environ.get("JOB_COMPLETION_INDEX")
    if job_index is not None:
        return job_index == "0"

    pet_rank = os.environ.get("PET_NODE_RANK")
    if pet_rank is not None:
        return pet_rank == "0"

    logger.debug(
        "is_primary_pod: neither JOB_COMPLETION_INDEX nor PET_NODE_RANK is set; returning False"
    )
    return False
```

**Review by briangallagher — COMMENTED** [review]
@Fiona-Waters Couple small points.

File: kubeflow/trainer/rhai/traininghub.py (hunk 5 above)
Comment (MStokluska) [reply kubeflow/trainer/rhai/traininghub.py]: BTW, I wonder if there's a better way of handling the resolution. This is now the second place where we will need to update the logic when adding new algorithms.

Comment (Fiona-Waters) [reply kubeflow/trainer/rhai/traininghub.py]: Yes you’re right @MStokluska let me look at fixing this.

File: kubeflow/trainer/rhai/traininghub.py (hunk 6 above)
Comment (Fiona-Waters) [reply kubeflow/trainer/rhai/traininghub.py]: I've added this function in this [commit](https://github.com/opendatahub-io/kubeflow-sdk/pull/66/changes/726d3a4e8130ddab5a05be67151e0a0cead37b37) please take a look

File: kubeflow/trainer/rhai/traininghub.py (hunk 5 above)
Comment (Fiona-Waters) [reply kubeflow/trainer/rhai/traininghub.py]: @briangallagher @MStokluska I've added code that introduces a centralized algorithm registry to eliminate scattered algorithm-specific branching throughout the codebase, making it easier to add new training algorithms. Please take a look. https://github.com/opendatahub-io/kubeflow-sdk/pull/66/changes/cd2de1606624566490356ce8cecec0344b2e6c35 https://github.com/opendatahub-io/kubeflow-sdk/pull/66/changes/45f8c5c617b22341b5d41c291d4d82f03743ffb0
It might be a better option to move this to another PR and get the stale metrics cleanup included in this weeks release. Please let me know what you think.

**Review by MStokluska — APPROVED** [review]
Thanks Fiona, looks good to me!
